docs/software-architecture-and-design.md

```diff
@@ -167,40 +161,40 @@ Examples of this can be found in many places, including:
 
-Read [https://www.tutorialspoint.com/Explain-Cplusplus-Singleton-design-pattern] for more information.
+Read https://refactoring.guru/design-patterns/singleton for more information.
 
 We use the Singleton pattern for our logger. This allows us to create a single logger for the entire system, and code can make calls to the logger from anywhere, rather than us having to pass a `logger` object literally everywhere.
 
 
-## Factory Design Pattern
-The Factory Design Pattern is useful for hiding or abstracting how certain objects are created.
+## Factory Pattern
+The Factory pattern is useful for hiding or abstracting how certain objects are created.
 
-Read [https://www.geeksforgeeks.org/design-patterns-set-2-factory-method/] for more information.
+Read the Refactoring Guru articles on the [Factory Method pattern](https://refactoring.guru/design-patterns/factory-method) and the [Abstract Factory pattern](https://refactoring.guru/design-patterns/abstract-factory) for more information.
 
 Because the Factory needs to know about what objects are available to be created, it can be taken one step further to auto-register these object types. Rather than a developer having to remember to add code to the Factory every time they create a new class, this can be done "automatically" with some clever code. This helps reduce mistakes and saves developers work.
 
-Read [http://derydoca.com/2019/03/c-tutorial-auto-registering-factory/] for more information.
+Read http://derydoca.com/2019/03/c-tutorial-auto-registering-factory/ for more information.
 
 The auto-registering factory is particularly useful for our `PlayFactory`, which is responsible for creating [Plays](#plays). Every time we run our [AI](#ai) we want to know what [Plays](#plays) are available to choose from. The Factory pattern makes this really easy, and saves us having to remember to update some list of "available Plays" each time we add or remove one.
 
 The Factory pattern is also used to create different [Backends](#backend)
 
 
-## Visitor Design Pattern
-The `Visitor Design Pattern` is arguably the most "advanced" design pattern we use. It is used when we need to perform different operations on a group of "similar" objects, for example a bunch of objects that inherit from the same parent class ([Intents](#intents)). We might only know all these objects are an [Intent](#intent), but we don't know specifically which type each one is (eg. `MoveIntent` vs `KickIntent`). The Visitor Pattern helps us "recover" that type information so we can perform different operations on the different types of objects. It is generally preferred to a big `if-block` with a case for each type, because the compiler can help warn you when you've forgotten to handle a certain type, and therefore helps prevent mistakes.
+## Visitor Pattern
```

Comment:
Maybe we should switch the example form Intent to MotionConstraints

Reply:
done